### ISSUE TITLE
Issues with OIDC auth in 4.x with Azure

### ISSUE THREAD

**Issue by ralekseenkov** [concourse/concourse#2538]
# Bug Report

* Concourse version: 4.0.0
* Deployment type: Helm chart with the latest pull request for 4.0.0 changes
* Infrastructure/IaaS: Kubernetes
* Browser (if applicable): Chrome on Mac OS

We are trying to migrate from 3.14.1 to 4.0.0, preserving Azure-based Oauth and making good use of user groups. 

Right now my understanding is that, when using OIDC connector, there is no way to automatically retrieve user groups and tie them to Concourse teams. Mostly because OIDC connector coreos/dex doesn't support it - https://github.com/coreos/dex/issues/1065

When happens when you log in is that the list of groups is always empty:
```
{"timestamp":"1535500679.667008162","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"login successful: connector \"oidc\", name=\"Roman Alekseenkov\", email=\" (unverified)\", groups=[]","session":"6"}}
```

So, basically, when creating a new team there is no way to tie it to a certain group, as `--oidc-group` won't do anything because the list of groups for any user is always empty.

If my understanding is correct, this essentially means that:
1) The only 2 connectors with decent group support right now are GitHub and CF
2) When authenticating via Azure, Google Cloud, etc you will be forced to call set-team and manually configure the list of users. For our use case with a lot of users this is a big challenge.

What I'd like to understand is:
1) What's the answer for a larger community (non-github auth, non-cf/non-bosh auth) who want to have good RBAC support on top of Azure or Google Cloud auth
2) Are there any plans to add support for other connectors which have good support for groups? E.g. SAML 2.0 or LDAP

If my assessment is incorrect and we can make groups work with Azure or GKE, please let me know.

**Comment by danilopopeye** [comment]
> Right now my understanding is that, when using OIDC connector, there is no way to automatically retrieve user groups and tie them to Concourse teams.

I'm using the OIDC groups with Okta with these settings:

```
CONCOURSE_OIDC_SCOPE='openid email profile address phone offline_access groups'
CONCOURSE_MAIN_TEAM_OIDC_GROUP='<gorup-admin>'
```

The auth log:

```
{"timestamp":"1534768963.507692099","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"login successful: connector \"oidc\", name=\"Danilo\", email=\"danilo@XXXX (unverified)\", groups=[\"admin\" \"concourse\"]","session":"5"}}
```

**Comment by danilopopeye** [comment]
> When authenticating via Azure, Google Cloud, etc you will be forced to call set-team and manually configure the list of users. For our use case with a lot of users this is a big challenge.

yeah, this is something I haven't figured out yet. The creation of groups is manually done and referenced with the `oidc:<group>` via `fly set-team`

**Comment by danilopopeye** [comment]
> The next step for me will be to try OKTA as well.

@ralekseenkov no problem. let me known if you run into any trouble configuring Okta 😉

**Comment by pivotal-jwinters** [comment]
@ralekseenkov the oidc connector will support mapping groups from the 3rd party token, but it doesn't support making a separate API request for groups. By default we search for a claim called `groups` (I'm guessing this is why okta works?).

Does Azure return groups in the token? If so you can map it using the `oidc-groups-key` flag.

**Comment by ralekseenkov** [comment]
@pivotal-jwinters yes, that's exactly what we tried, mapping groups to a claim field with `oidc-groups-key`, but couldn't get it to work with Azure endpoints.

v2 don't seem to support "groups" at all - https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-limitations

v1 it's supposed to work with it, but there are some caveats. E.g.:
```
To ensure that the token size doesn’t exceed HTTP header size limits, Azure AD limits the
number of objectIds that it includes in the groups claim. If a user is member of more groups
than the overage limit, then Azure AD does not emit the groups claim in the token. Instead, it
includes an overage claim in the token that indicates to the application to query the Graph API
to retrieve the user’s group membership.
```

So we spent a day trying to make this work in our setup and it didn't. Then, on top of all Azure issues we saw that OIDC connector doesn't support groups and then we gave up on it. My plan is still remains to try integration with OKTA and we are waiting on IT to create an app.

Another 2c... I would recommend that you guys update README.md in your fork of dex and say that OIDC support for groups = 'yes' (due to https://github.com/concourse/dex/commit/94cb3dabcceef6f3c3725fe3df02d922f311bc61). Because right now it says 'no' on the front page, and this is something that contributed to the issue and led us in the wrong direction.

**Comment by pivotal-jwinters** [comment]
@ralekseenkov I've update the README in our dex fork. Thanks for pointing it out.

Glad you have a path forward with Okta. Just an FYI we do also support LDAP. And adding support for SAML would really just be a matter of proxying some config. I should be easy for someone to PR (wink wink) if they have an existing SAML service to test against.

**Comment by ralekseenkov** [comment]
FYI - for people who'd want to set Concourse up with Azure v1 endpoinds... we got it somewhat working.

It required us to edit an magic property in registered app manifest:
`"groupMembershipClaims": "All"`

After that it started returning groups correctly. Username still doesn't show up correctly though
```
"message":"login successful: connector \"oidc\", username=\"\", email=\" (unverified)\", groups=[\"835a7a08-0f2e-....\" \"ab699d36-6085-....\"
```

<img width="1269" alt="screen shot 2018-09-21 at 1 56 00 pm" src="https://user-images.githubusercontent.com/609325/45905710-1d624380-bda6-11e8-9799-73d157a30d69.png">

**Comment by ralekseenkov** [comment]
Magically started working with Azure v2 endpoints and the following config for the Helm chart: 
```
  oidcAuth:
    enabled: true
    issuer: "https://login.microsoftonline.com/xxx/v2.0"
    clientId: "xxx"
    clientSecret: "xxx"
    scope: "openid,profile,email"
```

May be it's going to be helpful to someone. Closing this, thanks for the help everyone

**Comment by mattronix** [comment]
@ralekseenkov How did you configure your app registration? 

Im using the same config and find that the username attribute is blank and we only get the users email.

Our users dont get a long hash so im not sure how to set permissions for them.

**Comment by mattronix** [comment]
the users can login via our OIDC connection to ADFS but username shows:

{"timestamp":"1544182085.880118370","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"login successful: connector \"oidc\", username=\"\", email=\"Matthew.Frost@domain.tld (unverified)\", groups=[]","session":"5"}}

**Comment by ralekseenkov** [comment]
@mattronix we do permission based on groups. in order for Azure AD to return security groups for every used, you need to need to do:

App Permissions: Sign in and read user profile
App Manifest: “groupMembershipClaims”: “SecurityGroup”

Hope this helps

**Comment by mattronix** [comment]
@ralekseenkov  Awesome this fixed it for me!

**Comment by mattronix** [comment]
However i did “groupMembershipClaims”: “All”

**Comment by ralekseenkov** [comment]
@mattronix I would not recommend “all”.

We used it initially and had to go to “SecurityGroup”. What happens with “All” is once you have a user with too many security groups and inter groups, Azure will return an empty list of groups and expect the client (Concourse) to retrieve them via a separate call. Concourse doesn’t support a separate call, so auth will fail in a sense that user will see empty set of pipelines in Concourse. So it will work for some users, but will fail for users belonging to many groups.

Using “SecurityGroup” cut down the list of returned values by 80% for us

**Comment by gbvanrenswoude** [comment]
@danilopopeye how did you do the OKTA side of things (app setup)?

**Comment by danilopopeye** [comment]
> @danilopopeye how did you do the OKTA side of things (app setup)?

@gbvanrenswoude see if this helps 😃 

![concourse-okta-config](https://user-images.githubusercontent.com/27297/59766625-fd00e000-9276-11e9-8c17-cafd6944d8cd.jpg)

```
CONCOURSE_OIDC_DISPLAY_NAME="auth name on login page"
CONCOURSE_OIDC_ISSUER=https://my-instance.okta.com/oauth2/default
CONCOURSE_OIDC_CLIENT_ID=0oaba5l2oxzFG9Zci356
CONCOURSE_OIDC_CLIENT_SECRET=secret
CONCOURSE_OIDC_SCOPE='openid email profile address phone offline_access groups'
CONCOURSE_MAIN_TEAM_OIDC_GROUP='some_group,some_other_group'
```

**Comment by gbvanrenswoude** [comment]
@danilopopeye  Many thanks! I replicated your setup but the groups stay empty:
```
{"timestamp":"2019-06-19T15:13:45.834970704Z","level":"info","source":"atc","message":"atc.dex.event","data":{"fields":{},"message":"login successful: connector \"oidc\", username=\"\", email=\"me@corp.com (unverified)\", groups=[]","session":"7"}}
{"timestamp":"2019-06-19T15:13:45.914043028Z","level":"error","source":"atc","message":"atc.sky.callback.failed-to-issue-concourse-token","data":{"error":"user doesn't belong to any team","session":"6.10"}
```

Did you specifically do something to get the groups to work?
